crates/astria-sequencer/src/ibc/ics20_transfer.rs

```diff
@@ -288,290 +353,268 @@ impl AppHandlerExecute for Ics20Transfer {
-                ))?,
+    // Since we are refunding tokens, packet_data.sender is an address on Astria:
+    // the packet was not commited on the counter party chain but returned as-is.
+    let receiver = parse_sender(&state, &packet_data.sender)
+        .await
+        .with_context(|| {
+            format!(
+                "failed parsing packet.sender `{}` as the return address",
+                packet_data.sender
+            )
+        })?;
+
+    let asset = parse_asset(&state, &packet_data.denom)
+        .await
+        .with_context(|| format!("failed parsing packet.asset `{}`", packet_data.denom))?;
+
+    if let Ok(memo) = serde_json::from_str::<Ics20WithdrawalFromRollup>(&packet_data.memo) {
+        info!(
+                memo.rollup_return_address,
+                memo_name = astria_core::generated::protocol::memos::v1alpha1::Ics20WithdrawalFromRollup::full_name(),
+                "was able to parse memo; refunding to rollup",
+            );
+        refund_tokens_to_rollup(&mut state, receiver, asset, amount, memo)
+            .await
+            .context("failed to refund the rollup")?;
+    } else {
+        info!(
+            recipient = packet_data.sender,
+            "memo in fungible token packet was unknown; taking packet sender as recipient address \
+             on sequencer",
+        );
+        let recipient = packet_data.sender.parse().with_context(|| {
+            format!(
+                "failed to parse recipient `{}` as Astria address",
+                packet_data.sender
             )
-            .context("failed to update escrow account balance in execute_ics20_transfer")?;
+        })?;
+        refund_tokens_to_sequencer(
+            &mut state,
+            recipient,
+            asset,
+            amount,
+            &packet.port_on_a,
+            &packet.chan_on_a,
+        )
+        .await
+        .context("failed to refund a sequencer address")?;
+    }
 
+    Ok(())
+}
+#[instrument(skip_all)]
+async fn refund_tokens_to_sequencer<S: StateWrite>(
+    mut state: S,
+    recipient: Address,
+    asset: denom::TracePrefixed,
+    amount: u128,
+    source_port: &PortId,
+    source_channel: &ChannelId,
+) -> Result<()> {
+    if is_refund_source_zone(&asset, source_port, source_channel) {
         state
-            .increase_balance(recipient, &denom_trace, packet_amount)
+            .decrease_ibc_channel_balance(source_channel, &asset, amount)
+            .await
+            .context("failed to withdraw refund amount from ibc20 channel")?;
+        state
+            .increase_balance(recipient, &asset, amount)
             .await
             .context("failed to update user account balance in execute_ics20_transfer")?;
     } else {
         // register denomination in global ID -> denom map if it's not already there
+        // XXX(janis): is this correct? Shouldn't an asset that is being refunded always
+        // be present in the state if one attempted to transfer it out?
```

Comment:
yes, you can remove this, if the asset was withdrawn it must have existed in the state already (as it's either native or was transferred in)

Reply:
Nice! Good that my intuition was correct. Done in  69a86d02